[Rector] Add ImplementsFactoryInterfaceToPsrFactoryRector (#35)
laminas/laminas-servicemanager-migration: docs/book/v1/migrate-implements-factory-interface-to-use-psr-container.md

To migrate existing factory classes that implements `Laminas\ServiceManager\Factory\FactoryInterface`, there is a Rector rule that can be used to migrate for that. For example, we have the following factory class:

```php
use My\Service;
use Interop\Container\ContainerInterface;
use Laminas\ServiceManager\Factory\FactoryInterface;

class ServiceFactory implements FactoryInterface
{
    public function __invoke(ContainerInterface $container, $requestedName, array $options = null)
    {
        return new Service();
    }
}
```

We need to migrate to :

```php
use My\Service;
use Interop\Container\ContainerInterface;
use Laminas\ServiceManager\Factory\FactoryInterface;

class ServiceFactory
{
    public function __invoke(\Psr\Container\ContainerInterface $container)
    {
        return new Service();
    }
}
```

The steps to apply the changes are:

- Remove the interface from Factory class
- Replace `invoke__()` class method param and the use of Interop class in the class

To apply that, we can register rector rule: `Laminas\ServiceManager\Migration\Rector\Class_\ImplementsFactoryInterfaceToPsrFactoryRector` to our `rector.php` as an individual service:

> ```php

Reviewer: We can point out important lines:

```suggestion
<!-- markdownlint-disable MD033 -->
<pre class="language-php" data-line="11-15"><code>
```

use Rector\Core\Configuration\Option;
use Laminas\ServiceManager\Migration\Rector\Set\ValueObject\SetList;
use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;

return static function (ContainerConfigurator $containerConfigurator): void {
    $containerConfigurator->import(SetList::LAMINAS_SERVICEMANGER_40);

    $parameters = $containerConfigurator->parameters();
    $parameters->set(Option::PATHS, [__DIR__ . '/module']);

     // register ImplementsFactoryInterfaceToPsrFactoryRector service
    $services = $containerConfigurator->services();
    $services->set(
        \Laminas\ServiceManager\Migration\Rector\Class_\ImplementsFactoryInterfaceToPsrFactoryRector::class
    );
};

> ```

Reviewer: ```suggestion
</code></pre>
<!-- markdownlint-restore -->
```

If we want to auto import, we can use `LAMINAS_SERVICEMANGER_40_AUTO_IMPORT`:

> ```php

Reviewer: ```suggestion
<!-- markdownlint-disable MD033 -->
<pre class="language-php" data-line="6"><code>
```

use Rector\Core\Configuration\Option;
use Laminas\ServiceManager\Migration\Rector\Set\ValueObject\SetList;
use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;

return static function (ContainerConfigurator $containerConfigurator): void {
    $containerConfigurator->import(SetList::LAMINAS_SERVICEMANGER_40_AUTO_IMPORT);

    $parameters = $containerConfigurator->parameters();
    $parameters->set(Option::PATHS, [__DIR__ . '/module']);

    // register ImplementsFactoryInterfaceToPsrFactoryRector service
    $services = $containerConfigurator->services();
    $services->set(
        \Laminas\ServiceManager\Migration\Rector\Class_\ImplementsFactoryInterfaceToPsrFactoryRector::class
    );
};

> ```

Reviewer: ```suggestion
</code></pre>
<!-- markdownlint-restore -->
```

After configuration in place, we can run:

```bash
vendor/bin/rector process --dry-run
```

Ensure that the change is correct, if everything ok, we can run the fix:

```bash
vendor/bin/rector process
```

## Additional Adjustment

- Add Return type for the service, for example:

```php
use My\Service;
use Psr\Container\ContainerInterface;
use Laminas\ServiceManager\Factory\FactoryInterface;

class ServiceFactory
{
    public function __invoke(ContainerInterface $container): Service
    {
        return new Service();
    }
}
```

> For add return type by new instance creation, we can use [`ReturnTypeFromReturnNewRector` TypeDeclaration rule](https://github.com/rectorphp/rector/blob/main/docs/rector_rules_overview.md#returntypefromreturnnewrector).

Reviewer: A short code example like above would be great.

Author: updated code before after like above https://github.com/laminas/laminas-servicemanager-migration/pull/35/commits/3c0c1caa1b3af4ab5d97078da8e9bd11ca828c12

    );
};
</code></pre>
<!-- markdownlint-restore -->

After configuration in place, we can run:

```bash
vendor/bin/rector process --dry-run
```

Ensure that the change is correct, if everything ok, we can run the fix:

```bash
vendor/bin/rector process
```

## Additional Adjustment

- Add Return type for the service, for example, from the following:

<!-- markdownlint-disable MD033 -->
<pre class="language-php" data-line="7"><code>
use My\Service;
use Psr\Container\ContainerInterface;
use Laminas\ServiceManager\Factory\FactoryInterface;

class ServiceFactory
{
    public function __invoke(ContainerInterface $container)
    {
        return new Service();
    }
}
</code></pre>
<!-- markdownlint-restore -->

to have return type:

<!-- markdownlint-disable MD033 -->
<pre class="language-php" data-line="7"><code>
use My\Service;
use Psr\Container\ContainerInterface;
use Laminas\ServiceManager\Factory\FactoryInterface;

class ServiceFactory
{
    public function __invoke(ContainerInterface $container): Service
    {
        return new Service();
    }
}
</code></pre>
<!-- markdownlint-restore -->

> To add return type by new instance creation like above, we can use [`ReturnTypeFromReturnNewRector` TypeDeclaration rule](https://github.com/rectorphp/rector/blob/main/docs/rector_rules_overview.md#returntypefromreturnnewrector).

Reviewer: ```suggestion
### Add Return Type to Factory

To add return type by new instance creation, the standard Rector rule [`ReturnTypeFromReturnNewRector`](https://github.com/rectorphp/rector/blob/main/docs/rector_rules_overview.md#returntypefromreturnnewrector) can be used.

Register the rule in `rector.php`:

<!-- markdownlint-disable MD033 -->
<pre class="language-php" data-line="15"><code>
use Laminas\ServiceManager\Migration\Rector\Class_\ImplementsFactoryInterfaceToPsrFactoryRector;
use Laminas\ServiceManager\Migration\Rector\Set\ValueObject\SetList;
use Rector\Core\Configuration\Option;
use Rector\TypeDeclaration\Rector\ClassMethod\ReturnTypeFromReturnNewRector;
use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;

return static function (ContainerConfigurator $containerConfigurator): void {
    $containerConfigurator->import(SetList::LAMINAS_SERVICEMANGER_40_AUTO_IMPORT);

    $parameters = $containerConfigurator->parameters();
    $parameters->set(Option::PATHS, [__DIR__ . '/module']);

    $services = $containerConfigurator->services();
    $services->set(ImplementsFactoryInterfaceToPsrFactoryRector::class);
    $services->set(ReturnTypeFromReturnNewRector::class);
};
</code></pre>
<!-- markdownlint-restore -->

Result:

<!-- markdownlint-disable MD033 -->
<pre class="language-php" data-line="7"><code>
use My\Service;
use Psr\Container\ContainerInterface;
use Laminas\ServiceManager\Factory\FactoryInterface;

class ServiceFactory
{
    public function __invoke(ContainerInterface $container): Service
    {
        return new Service();
    }
}
</code></pre>
<!-- markdownlint-restore -->

Author: updated https://github.com/laminas/laminas-servicemanager-migration/pull/35/commits/74c26ff5ded5437053aca0639af02038f9ae86cf

use Laminas\ServiceManager\Migration\Rector\Set\ValueObject\SetList;
use Rector\Core\Configuration\Option;
use Rector\TypeDeclaration\Rector\ClassMethod\ReturnTypeFromReturnNewRector;
use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;

return static function (ContainerConfigurator $containerConfigurator): void {
    $containerConfigurator->import(SetList::LAMINAS_SERVICEMANGER_40_AUTO_IMPORT);

    $parameters = $containerConfigurator->parameters();
    $parameters->set(Option::PATHS, [__DIR__ . '/module']);

    $services = $containerConfigurator->services();
    $services->set(ImplementsFactoryInterfaceToPsrFactoryRector::class);
    $services->set(ReturnTypeFromReturnNewRector::class);
};
</code></pre>
<!-- markdownlint-restore -->

Result:

<!-- markdownlint-disable MD033 -->
<pre class="language-php" data-line="7"><code>
use My\Service;
use Psr\Container\ContainerInterface;
use Laminas\ServiceManager\Factory\FactoryInterface;

class ServiceFactory
{
    public function __invoke(ContainerInterface $container): Service
    {
        return new Service();
    }
}
</code></pre>
<!-- markdownlint-restore -->

> ### Sort use statement
> 
> To sort it, we may require coding style tool for it, for example: [PHP-CS-Fixer](https://github.com/FriendsOfPHP/PHP-CS-Fixer) or [PHP_CodeSniffer](https://github.com/squizlabs/PHP_CodeSniffer)

Reviewer: ```suggestion
### Sort Use Statements

To sort the `use` statements for the factory a coding style tool can be used, for example: [PHP_CodeSniffer](https://github.com/squizlabs/PHP_CodeSniffer) with [Slevomat Coding Standard](https://github.com/slevomat/coding-standard#slevomatcodingstandardnamespacesalphabeticallysorteduses-).
```

Author: updated https://github.com/laminas/laminas-servicemanager-migration/pull/35/commits/cffa3497e111e67fb3c0a41d5248a667aa3606de

# Migrate Class Implements FactoryInterface to use Psr Container

To migrate existing factory classes that implements `Laminas\ServiceManager\Factory\FactoryInterface`, there is a Rector rule that can be used to migrate for that. For example, we have the following factory class:

```php
use My\Service;
use Interop\Container\ContainerInterface;
use Laminas\ServiceManager\Factory\FactoryInterface;

> class ImplementsRootAbstractFactoryInterface implements FactoryInterface

Reviewer: ```suggestion
class ServiceFactory implements FactoryInterface
```

{
    public function __invoke(ContainerInterface $container, $requestedName, array $options = null)
    {
        return new Service();
    }
}
```

We need to migrate to :

```php
use My\Service;
use Interop\Container\ContainerInterface;
use Laminas\ServiceManager\Factory\FactoryInterface;

> class ImplementsRootAbstractFactoryInterface

Reviewer: ```suggestion
class ServiceFactory
```

Author: updated https://github.com/laminas/laminas-servicemanager-migration/pull/35/commits/dea08254d190d269625e034ddef2a0ec7eeb3757

{
    public function __invoke(\Psr\Container\ContainerInterface $container)
    {
        return new Service();
    }
}

> ```

Reviewer: This is not the real goal, only an intermediate step. 

The goal should be:

```php
use My\Service;
use Laminas\ServiceManager\Factory\FactoryInterface;
use Psr\Container\ContainerInterface;

class ServiceFactory
{
    public function __invoke(ContainerInterface $container)
    {
        return new Service();
    }
}
```

Better:

```php
use Laminas\ServiceManager\Factory\FactoryInterface;
use My\Service;
use Psr\Container\ContainerInterface;

final class ServiceFactory
{
    public function __invoke(ContainerInterface $container): Service
    {
        return new Service();
    }
}
```

We should describe all the steps to achieve this goal in order to show the whole workflow. (All used standard Rector rules should be linked with the original documentation.)

Suggestion for the steps:

1. Remove the Interface from the Factory
1. Additional Adjustments
   1. Auto Import Names
   2. Finalize Factory Class
   3. Add Return Type

And add a note on how to sort the use statements via CS tools.

Author: @froschdesign add return type hint is not the original issue you wrote at https://github.com/laminas/laminas-servicemanager-migration/issues/34

There is TypeDeclaration rector rule for that, but out of this documentation scope, possibly can be noted to link, but not mandatory.

https://github.com/rectorphp/rector/blob/main/docs/rector_rules_overview.md#typedeclaration

Reviewer: > There is TypeDeclaration rector rule for that, but out of this documentation scope, possibly can be noted to link, but not mandatory.

Right, that's a standard rule of Rector, but that doesn't mean we have to show only half an example in documentation. Therefore I suggested the headline "Additional Adjustments".
And if a standard rule is used then also a link must be added.

Author: Ok, updated with add step + additional adjustment https://github.com/laminas/laminas-servicemanager-migration/pull/35/commits/544547babdc7787d9cc1333a0f3cddbe87885089

What we need to do is register rector rule: `Laminas\ServiceManager\Migration\Rector\Class_\ImplementsFactoryInterfaceToPsrFactoryRector` to our `rector.php` as an individual service:

```php

> <?php
> // rector.php
> declare(strict_types=1);
> 

Reviewer: ```suggestion
```

use Rector\Core\Configuration\Option;
use Laminas\ServiceManager\Migration\Rector\Set\ValueObject\SetList;
use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;

return static function (ContainerConfigurator $containerConfigurator): void {
    $containerConfigurator->import(SetList::LAMINAS_SERVICEMANGER_40);

    $parameters = $containerConfigurator->parameters();
    $parameters->set(Option::PATHS, [__DIR__ . '/module']);

     // register ImplementsFactoryInterfaceToPsrFactoryRector service
    $services = $containerConfigurator->services();
    $services->set(
        \Laminas\ServiceManager\Migration\Rector\Class_\ImplementsFactoryInterfaceToPsrFactoryRector::class
    );
};
```

If we want to auto import, we can use `LAMINAS_SERVICEMANGER_40_AUTO_IMPORT`:

```php

> <?php
> // rector.php
> declare(strict_types=1);

Reviewer: ```suggestion
```

Author: removed https://github.com/laminas/laminas-servicemanager-migration/pull/35/commits/02df9817ab575cba291b8ebb1b88b4fa2829838c
